src/main/java/com/salesforce/dynamodbv2/mt/util/CachingAmazonDynamoDbStreams.java

```diff
@@ -529,6 +568,97 @@ private CachingAmazonDynamoDbStreams(AmazonDynamoDBStreams amazonDynamoDbStreams
+     * @param describeStreamRequest Describe stream request.
+     * @return Stream details for the given request (streamArn).
+     */
+    private DescribeStreamResult lookupDescribeStreamResult(DescribeStreamRequest describeStreamRequest)
+        throws AmazonDynamoDBException {
+
+        String key = describeStreamRequest.getStreamArn();
+        try {
+            DescribeStreamResult result = describeStreamCache.get(key,
+                () -> this.loadStreamDescriptionForAllShards(describeStreamRequest));
+            return result;
+        } catch (UncheckedExecutionException | ExecutionException | InvalidCacheLoadException e) {
+            // Catch exceptions thrown on cache lookup or cache loader and try load method once more.
+            // (get call will throw UncheckedExecutionException before aws exception (i.e. AmazonDynamoDBException)).
+            LOG.warn("Failed getting DescribeStreamResult for all shards from cache lookup or load method.  "
+                + "Retrying describeStream call.");
+            e.printStackTrace();
+            return this.loadStreamDescriptionForAllShards(describeStreamRequest);
+        }
+    }
+
+    /**
+     * Gets the {@code DescribeStreamResult} from the DescribeStream API.
+     * @param describeStreamRequest Describe stream request.
+     * @return Stream details for the given request with all the shards currently available.
+     */
+    protected DescribeStreamResult loadStreamDescriptionForAllShards(DescribeStreamRequest describeStreamRequest)
+        throws AmazonDynamoDBException {
+
+        List<Shard> allShards = new ArrayList<>();
+        String streamArn = describeStreamRequest.getStreamArn();
+        String lastShardId = describeStreamRequest.getExclusiveStartShardId();
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("cache miss in describe stream cache for key: {}", streamArn);
+        }
+        // TODO metrics: count cache miss
+
+        DescribeStreamResult result = null;
+        DescribeStreamResult previousResult;
+        do {
+            describeStreamRequest = new DescribeStreamRequest().withStreamArn(streamArn)
+                .withExclusiveStartShardId(lastShardId);
+            previousResult = result;
+            result = super.describeStream(describeStreamRequest);
+
+            if (result != null && result.getStreamDescription().getShards() != null) {
+                List<Shard> shards = result.getStreamDescription().getShards();
+                allShards.addAll(shards);
+                lastShardId = result.getStreamDescription().getLastEvaluatedShardId();
+            } else {
+                // since the request was null, update lastShardId to null since we can't get the next request
+                // from a null response
+                lastShardId = null;
+            }
+        } while (lastShardId != null);
+
+        if (result != null) {
```

Comment:
Why not just create a new `DescribeStreamResult` at the end from whatever was loaded?

Reply:
After removing the null checks, I first changed it to this:
```
result.getStreamDescription().setShards(allShards);
result.setStreamDescription(result.getStreamDescription());
```

then to this
```
new DescribeStreamResult().withStreamDescription(result.getStreamDescription().withShards(allShards));
```

I think either way is fine.